README.md
# bip38

[![build status](https://secure.travis-ci.org/bitcoinjs/bip38.svg)](http://travis-ci.org/bitcoinjs/bip38)
[![Coverage Status](https://img.shields.io/coveralls/cryptocoinjs/bip38.svg)](https://coveralls.io/r/cryptocoinjs/bip38)
[![Version](http://img.shields.io/npm/v/bip38.svg)](https://www.npmjs.org/package/bip38)

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

A JavaScript component that adheres to the [BIP38](https://github.com/bitcoin/bips/blob/master/bip-0038.mediawiki) standard to secure your crypto currency private keys. Fully compliant with Node.js and the browser (via Browserify).


## Why?
BIP38 is a standard process to encrypt Bitcoin and crypto currency private keys that is impervious to brute force attacks thus protecting the user.


## Package Info
- homepage: [http://cryptocoinjs.com/modules/currency/bip38/](http://cryptocoinjs.com/modules/currency/bip38/)
- github: [https://github.com/cryptocoinjs/bip38](https://github.com/cryptocoinjs/bip38)
- tests: [https://github.com/cryptocoinjs/bip38/tree/master/test](https://github.com/cryptocoinjs/bip38/tree/master/test)
- issues: [https://github.com/cryptocoinjs/bip38/issues](https://github.com/cryptocoinjs/bip38/issues)
- license: **MIT**
- versioning: [http://semver-ftw.org](http://semver-ftw.org)


## Usage

### Installation

    npm install --save bip38


### Async methods

Async methods are available, but using them will be slower, but free up the event loop in intervals you choose.

For benchmark results, please see the section in the README of the [scryptsy](https://github.com/cryptocoinjs/scryptsy/tree/395c3b09b21e06ea4a6cc2933e046c0984a414c5#benchmarks) library. Increasing the interval will decrease the performance hit, but increase the span between event loop free ups (UI drawing etc.) promiseInterval is the last optional parameter after scryptParams. There is no recommendation currently, as the performance trade off is app specific.

### API
### encrypt(buffer, compressed, passphrase[, progressCallback, scryptParams])

``` javascript
var bip38 = require('bip38')
var wif = require('wif')

var myWifString = '5KN7MzqK5wt2TP1fQCYyHBtDrXdJuXbUzm4A9rKAteGu3Qi5CVR'
var decoded = wif.decode(myWifString)

var encryptedKey = bip38.encrypt(decoded.privateKey, decoded.compressed, 'TestingOneTwoThree')
console.log(encryptedKey)
// => '6PRVWUbkzzsbcVac2qwfssoUJAN1Xhrg6bNk8J7Nzm5H7kxEbn2Nh2ZoGg'
```


### decrypt(encryptedKey, passphrase[, progressCallback, scryptParams])

``` javascript
var bip38 = require('bip38')
var wif = require('wif')

var encryptedKey = '6PRVWUbkzzsbcVac2qwfssoUJAN1Xhrg6bNk8J7Nzm5H7kxEbn2Nh2ZoGg'
var decryptedKey = bip38.decrypt(encryptedKey, 'TestingOneTwoThree', function (status) {
  console.log(status.percent) // will print the percent every time current increases by 1000
})

console.log(wif.encode(0x80, decryptedKey.privateKey, decryptedKey.compressed))
// => '5KN7MzqK5wt2TP1fQCYyHBtDrXdJuXbUzm4A9rKAteGu3Qi5CVR'
```


# References
- https://github.com/bitcoin/bips/blob/master/bip-0038.mediawiki
- https://github.com/pointbiz/bitaddress.org/issues/56 (Safari 6.05 issue)
- https://github.com/casascius/Bitcoin-Address-Utility/tree/master/Model
- https://github.com/nomorecoin/python-bip38-testing/blob/master/bip38.py
- https://github.com/pointbiz/bitaddress.org/blob/master/src/ninja.key.js


CHANGELOG.md
3.1.1 / 2020-05-31
------------------
- Add safe-buffer as explicit dependency

3.1.0 / 2020-04-09
------------------
- Add async methods

3.0.0 / 2019-09-12
------------------
- Fixed backwards incompatible bug with passphrase NFC normalization

2.0.3 / 2019-05-22
------------------
- made compatible for Electron v4

2.0.2 / 2017-12-14
------------------
- use safe-buffer
- upgrade scryptsy

2.0.1 / 2017-04-20
------------------
- upgrade bs58check
- index: rm unused address parameter

2.0.0 / 2016-12-20
------------------
- removed class instantiation. Removed `coinstring` dep.

1.4.0 / 2015-11-03
------------------
- added `progressCallback`. See: https://github.com/bitcoinjs/bip38/pull/16

1.3.0 / 2015-06-04
------------------
- use `createHash` and `aes` directly. https://github.com/cryptocoinjs/bip38
- JavaScript Standard Style

1.2.0 / 2015-01-05
------------------
- removed dependency upon `aes` package since Browserify now supports aes [Daniel Cousens](https://github.com/cryptocoinjs/bip38/pull/6)
- removed `crypto-browserify` devDep and removed `browser` field from `package.json`; no longer necessary
- added method `verify()` [Daniel Cousens](https://github.com/cryptocoinjs/bip38/pull/7)

1.1.1 / 2014-09-19
------------------
- bugfix: enforce zero padding [Daniel Cousens](https://github.com/cryptocoinjs/bip38/commit/e73598d0fc1d1b3c04c132c34053e96bec6bd201)
- add MIT license to package.json

1.1.0 / 2014-07-11
------------------
- added methods `encryptRaw` and `decryptRaw` [Daniel Cousens](https://github.com/cryptocoinjs/bip38/pull/4)

1.0.0 / 2014-06-10
------------------
- upgraded `"scryptsy": "~0.2.0"` to `"scryptsy": "^1.0.0"`
- upgraded  `"bigi": "~0.2.0"` to `"bigi": "^1.2.0"`
- removed `ecurve-names` dep
- upgraded `"ecurve": "~0.3.0"` to `"ecurve": "^0.8.0"`
- removed semicolons per http://cryptocoinjs.com/about/contributing/#semicolons
- removed `crypto-hashing` dep
- removed `bs58` dep, added `coinstring` dep
- removed `terst` for `assert`
- added TravisCI
- added Coveralls
- added testling
- removed static level methods, must call `new`

0.1.0 / 2014-03-05
------------------
- added support to decrypt ECMultiplied keys, #1
- made constructor work without `new`
- upgraded deps `ecurve`, `ecurve-names`, and `scryptsy`

0.0.1 / 2014-02-28
------------------
- initial release


test/bip38.js
var assert = require('assert')
var Bip38 = require('../')
var fixtures = require('./fixtures')

/* global beforeEach, describe, it */

describe('bip38', function () {
  this.timeout(70000)

  var bip38
  beforeEach(function () {
    bip38 = new Bip38()
  })

  describe('decrypt', function () {
    fixtures.valid.forEach(function (f) {
      it('should decrypt ' + f.description, function () {
        assert.equal(bip38.decrypt(f.bip38, f.passphrase), f.wif)
      })
    })

    fixtures.invalid.decrypt.forEach(function (f) {
      it('should throw ' + f.description, function () {
        assert.throws(function () {
          bip38.decrypt(f.bip38, f.passphrase)
        }, new RegExp(f.description, 'i'))
      })
    })

    fixtures.invalid.verify.forEach(function (f) {
      it('should throw because ' + f.description, function () {
        assert.throws(function () {
          bip38.decrypt(f.base58, 'foobar')
        }, new RegExp(f.exception))
      })
    })
  })

  describe('encrypt', function () {
    fixtures.valid.forEach(function (f) {
      if (f.decryptOnly) return

      it('should encrypt ' + f.description, function () {
        assert.equal(bip38.encrypt(f.wif, f.passphrase, f.address), f.bip38)
      })
    })
  })

  describe('verify', function () {
    fixtures.valid.forEach(function (f) {
      it('should return true for ' + f.bip38, function () {
        assert(bip38.verify(f.bip38))
      })
    })

    fixtures.invalid.verify.forEach(function (f) {
      it('should return false for ' + f.description, function () {
        assert(!bip38.verify(f.base58))
      })
    })
  })
})


test/index.js
/* global describe, it */

var assert = require('assert')
var bip38 = require('../')
var bs58check = require('bs58check')
var fixtures = require('./fixtures')
var wif = require('wif')

function replaceUnicode (str) {
  var map = {
    '\\u03D2\\u0301\\u{0000}\\u{00010400}\\u{0001F4A9}': '\u03D2\u0301\u{0000}\u{00010400}\u{0001F4A9}'
  }
  if (map[str]) str = map[str]
  return str
}

describe('bip38', function () {
  this.timeout(200000)

  describe('decrypt', function () {
    fixtures.valid.forEach(function (f) {
      it('should decrypt ' + f.description, function () {
        var result = bip38.decrypt(f.bip38, replaceUnicode(f.passphrase))

        assert.strictEqual(wif.encode(0x80, result.privateKey, result.compressed), f.wif)
      })
    })

    fixtures.invalid.decrypt.forEach(function (f) {
      it('should throw ' + f.description, function () {
        assert.throws(function () {
          bip38.decrypt(f.bip38, f.passphrase)
        }, new RegExp(f.description, 'i'))
      })
    })

    fixtures.invalid.verify.forEach(function (f) {
      it('should throw because ' + f.description, function () {
        assert.throws(function () {
          bip38.decrypt(f.base58, 'foobar')
        }, new RegExp(f.exception))
      })
    })
  })

  describe('encrypt', function () {
    fixtures.valid.forEach(function (f) {
      if (f.decryptOnly) return

      it('should encrypt ' + f.description, function () {
        var buffer = bs58check.decode(f.wif)

        assert.strictEqual(bip38.encrypt(buffer.slice(1, 33), !!buffer[33], replaceUnicode(f.passphrase)), f.bip38)
      })
    })
  })

  describe('decryptAsync', function () {
    fixtures.valid.forEach(function (f) {
      it('should decrypt ' + f.description, async function () {
        var result = await bip38.decryptAsync(f.bip38, replaceUnicode(f.passphrase))

        assert.strictEqual(wif.encode(0x80, result.privateKey, result.compressed), f.wif)
      })
    })

    fixtures.invalid.decrypt.forEach(function (f) {
      it('should throw ' + f.description, async function () {
        assert.rejects(async function () {
          await bip38.decryptAsync(f.bip38, replaceUnicode(f.passphrase))
        }, new RegExp(f.description, 'i'))
      })
    })

    fixtures.invalid.verify.forEach(function (f) {
      it('should throw because ' + f.description, async function () {
        assert.rejects(async function () {
          await bip38.decryptAsync(f.base58, 'foobar')
        }, new RegExp(f.exception))
      })
    })
  })

  describe('encryptAsync', function () {
    fixtures.valid.forEach(function (f) {
      if (f.decryptOnly) return

      it('should encrypt ' + f.description, async function () {
        var buffer = bs58check.decode(f.wif)

        assert.strictEqual(await bip38.encryptAsync(buffer.slice(1, 33), !!buffer[33], replaceUnicode(f.passphrase)), f.bip38)
      })
    })
  })

  describe('verify', function () {
    fixtures.valid.forEach(function (f) {
      it('should return true for ' + f.bip38, function () {
        assert(bip38.verify(f.bip38))
      })
    })

    fixtures.invalid.verify.forEach(function (f) {
      it('should return false for ' + f.description, function () {
        assert(!bip38.verify(f.base58))
      })
    })
  })
})


index.js
var aes = require('browserify-aes')
var assert = require('assert')
var Buffer = require('safe-buffer').Buffer
var bs58check = require('bs58check')
var createHash = require('create-hash')
var scrypt = require('scryptsy')
var xor = require('buffer-xor/inplace')

var ecurve = require('ecurve')
var curve = ecurve.getCurveByName('secp256k1')

var BigInteger = require('bigi')

// constants
var SCRYPT_PARAMS = {
  N: 16384, // specified by BIP38
  r: 8,
  p: 8
}
var NULL = Buffer.alloc(0)

function hash160 (buffer) {
  var hash
  try {
    hash = createHash('rmd160')
  } catch (e) {
    hash = createHash('ripemd160')
  }
  return hash.update(
    createHash('sha256').update(buffer).digest()
  ).digest()
}

function hash256 (buffer) {
  return createHash('sha256').update(
    createHash('sha256').update(buffer).digest()
  ).digest()
}

function getAddress (d, compressed) {
  var Q = curve.G.multiply(d).getEncoded(compressed)
  var hash = hash160(Q)
  var payload = Buffer.allocUnsafe(21)
  payload.writeUInt8(0x00, 0) // XXX TODO FIXME bitcoin only??? damn you BIP38
  hash.copy(payload, 1)

  return bs58check.encode(payload)
}

function prepareEncryptRaw (buffer, compressed, passphrase, scryptParams) {
  if (buffer.length !== 32) throw new Error('Invalid private key length')

  var d = BigInteger.fromBuffer(buffer)
  var address = getAddress(d, compressed)
  var secret = Buffer.from(passphrase.normalize('NFC'), 'utf8')
  var salt = hash256(address).slice(0, 4)

  var N = scryptParams.N
  var r = scryptParams.r
  var p = scryptParams.p

  return {
    secret,
    salt,
    N,
    r,
    p
  }
}

function finishEncryptRaw (buffer, compressed, salt, scryptBuf) {
  var derivedHalf1 = scryptBuf.slice(0, 32)
  var derivedHalf2 = scryptBuf.slice(32, 64)

  var xorBuf = xor(derivedHalf1, buffer)
  var cipher = aes.createCipheriv('aes-256-ecb', derivedHalf2, NULL)
  cipher.setAutoPadding(false)
  cipher.end(xorBuf)

  var cipherText = cipher.read()

  // 0x01 | 0x42 | flagByte | salt (4) | cipherText (32)
  var result = Buffer.allocUnsafe(7 + 32)
  result.writeUInt8(0x01, 0)
  result.writeUInt8(0x42, 1)
  result.writeUInt8(compressed ? 0xe0 : 0xc0, 2)
  salt.copy(result, 3)
  cipherText.copy(result, 7)

  return result
}

async function encryptRawAsync (buffer, compressed, passphrase, progressCallback, scryptParams, promiseInterval) {
  scryptParams = scryptParams || SCRYPT_PARAMS
  const {
    secret,
    salt,
    N,
    r,
    p
  } = prepareEncryptRaw(buffer, compressed, passphrase, scryptParams)

  var scryptBuf = await scrypt.async(secret, salt, N, r, p, 64, progressCallback, promiseInterval)

  return finishEncryptRaw(buffer, compressed, salt, scryptBuf)
}

function encryptRaw (buffer, compressed, passphrase, progressCallback, scryptParams) {
  scryptParams = scryptParams || SCRYPT_PARAMS
  const {
    secret,
    salt,
    N,
    r,
    p
  } = prepareEncryptRaw(buffer, compressed, passphrase, scryptParams)

  var scryptBuf = scrypt(secret, salt, N, r, p, 64, progressCallback)

  return finishEncryptRaw(buffer, compressed, salt, scryptBuf)
}

async function encryptAsync (buffer, compressed, passphrase, progressCallback, scryptParams, promiseInterval) {
  return bs58check.encode(await encryptRawAsync(buffer, compressed, passphrase, progressCallback, scryptParams, promiseInterval))
}

function encrypt (buffer, compressed, passphrase, progressCallback, scryptParams) {
  return bs58check.encode(encryptRaw(buffer, compressed, passphrase, progressCallback, scryptParams))
}

function prepareDecryptRaw (buffer, progressCallback, scryptParams) {
  // 39 bytes: 2 bytes prefix, 37 bytes payload
  if (buffer.length !== 39) throw new Error('Invalid BIP38 data length')
  if (buffer.readUInt8(0) !== 0x01) throw new Error('Invalid BIP38 prefix')

  // check if BIP38 EC multiply
  var type = buffer.readUInt8(1)
  if (type === 0x43) return { decryptEC: true }
  if (type !== 0x42) throw new Error('Invalid BIP38 type')

  var flagByte = buffer.readUInt8(2)
  var compressed = flagByte === 0xe0
  if (!compressed && flagByte !== 0xc0) throw new Error('Invalid BIP38 compression flag')

  var N = scryptParams.N
  var r = scryptParams.r
  var p = scryptParams.p

  var salt = buffer.slice(3, 7)
  return {
    salt,
    compressed,
    N,
    r,
    p
  }
}

function finishDecryptRaw (buffer, salt, compressed, scryptBuf) {
  var derivedHalf1 = scryptBuf.slice(0, 32)
  var derivedHalf2 = scryptBuf.slice(32, 64)

  var privKeyBuf = buffer.slice(7, 7 + 32)
  var decipher = aes.createDecipheriv('aes-256-ecb', derivedHalf2, NULL)
  decipher.setAutoPadding(false)
  decipher.end(privKeyBuf)

  var plainText = decipher.read()
  var privateKey = xor(derivedHalf1, plainText)

  // verify salt matches address
  var d = BigInteger.fromBuffer(privateKey)
  var address = getAddress(d, compressed)
  var checksum = hash256(address).slice(0, 4)
  assert.deepStrictEqual(salt, checksum)

  return {
    privateKey: privateKey,
    compressed: compressed
  }
}

async function decryptRawAsync (buffer, passphrase, progressCallback, scryptParams, promiseInterval) {
  scryptParams = scryptParams || SCRYPT_PARAMS
  const {
    salt,
    compressed,
    N,
    r,
    p,
    decryptEC
  } = prepareDecryptRaw(buffer, progressCallback, scryptParams)
  if (decryptEC === true) return decryptECMultAsync(buffer, passphrase, progressCallback, scryptParams, promiseInterval)

  var scryptBuf = await scrypt.async(passphrase.normalize('NFC'), salt, N, r, p, 64, progressCallback, promiseInterval)
  return finishDecryptRaw(buffer, salt, compressed, scryptBuf)
}

// some of the techniques borrowed from: https://github.com/pointbiz/bitaddress.org
function decryptRaw (buffer, passphrase, progressCallback, scryptParams) {
  scryptParams = scryptParams || SCRYPT_PARAMS
  const {
    salt,
    compressed,
    N,
    r,
    p,
    decryptEC
  } = prepareDecryptRaw(buffer, progressCallback, scryptParams)
  if (decryptEC === true) return decryptECMult(buffer, passphrase, progressCallback, scryptParams)
  var scryptBuf = scrypt(passphrase.normalize('NFC'), salt, N, r, p, 64, progressCallback)
  return finishDecryptRaw(buffer, salt, compressed, scryptBuf)
}

async function decryptAsync (string, passphrase, progressCallback, scryptParams, promiseInterval) {
  return decryptRawAsync(bs58check.decode(string), passphrase, progressCallback, scryptParams, promiseInterval)
}

function decrypt (string, passphrase, progressCallback, scryptParams) {
  return decryptRaw(bs58check.decode(string), passphrase, progressCallback, scryptParams)
}

function prepareDecryptECMult (buffer, passphrase, progressCallback, scryptParams) {
  var flag = buffer.readUInt8(1)
  var compressed = (flag & 0x20) !== 0
  var hasLotSeq = (flag & 0x04) !== 0

  assert.strictEqual((flag & 0x24), flag, 'Invalid private key.')

  var addressHash = buffer.slice(2, 6)
  var ownerEntropy = buffer.slice(6, 14)
  var ownerSalt

  // 4 bytes ownerSalt if 4 bytes lot/sequence
  if (hasLotSeq) {
    ownerSalt = ownerEntropy.slice(0, 4)

  // else, 8 bytes ownerSalt
  } else {
    ownerSalt = ownerEntropy
  }

  var encryptedPart1 = buffer.slice(14, 22) // First 8 bytes
  var encryptedPart2 = buffer.slice(22, 38) // 16 bytes

  var N = scryptParams.N
  var r = scryptParams.r
  var p = scryptParams.p
  return {
    addressHash,
    encryptedPart1,
    encryptedPart2,
    ownerEntropy,
    ownerSalt,
    hasLotSeq,
    compressed,
    N,
    r,
    p
  }
}

function getPassIntAndPoint (preFactor, ownerEntropy, hasLotSeq) {
  var passFactor
  if (hasLotSeq) {
    var hashTarget = Buffer.concat([preFactor, ownerEntropy])
    passFactor = hash256(hashTarget)
  } else {
    passFactor = preFactor
  }
  const passInt = BigInteger.fromBuffer(passFactor)
  return {
    passInt,
    passPoint: curve.G.multiply(passInt).getEncoded(true)
  }
}

function finishDecryptECMult (seedBPass, encryptedPart1, encryptedPart2, passInt, compressed) {
  var derivedHalf1 = seedBPass.slice(0, 32)
  var derivedHalf2 = seedBPass.slice(32, 64)

  var decipher = aes.createDecipheriv('aes-256-ecb', derivedHalf2, Buffer.alloc(0))
  decipher.setAutoPadding(false)
  decipher.end(encryptedPart2)

  var decryptedPart2 = decipher.read()
  var tmp = xor(decryptedPart2, derivedHalf1.slice(16, 32))
  var seedBPart2 = tmp.slice(8, 16)

  var decipher2 = aes.createDecipheriv('aes-256-ecb', derivedHalf2, Buffer.alloc(0))
  decipher2.setAutoPadding(false)
  decipher2.write(encryptedPart1) // first 8 bytes
  decipher2.end(tmp.slice(0, 8)) // last 8 bytes

  var seedBPart1 = xor(decipher2.read(), derivedHalf1.slice(0, 16))
  var seedB = Buffer.concat([seedBPart1, seedBPart2], 24)
  var factorB = BigInteger.fromBuffer(hash256(seedB))

  // d = passFactor * factorB (mod n)
  var d = passInt.multiply(factorB).mod(curve.n)

  return {
    privateKey: d.toBuffer(32),
    compressed: compressed
  }
}

async function decryptECMultAsync (buffer, passphrase, progressCallback, scryptParams, promiseInterval) {
  buffer = buffer.slice(1) // FIXME: we can avoid this
  passphrase = Buffer.from(passphrase.normalize('NFC'), 'utf8')
  scryptParams = scryptParams || SCRYPT_PARAMS
  const {
    addressHash,
    encryptedPart1,
    encryptedPart2,
    ownerEntropy,
    ownerSalt,
    hasLotSeq,
    compressed,
    N,
    r,
    p
  } = prepareDecryptECMult(buffer, passphrase, progressCallback, scryptParams)

  var preFactor = await scrypt.async(passphrase, ownerSalt, N, r, p, 32, progressCallback, promiseInterval)

  const {
    passInt,
    passPoint
  } = getPassIntAndPoint(preFactor, ownerEntropy, hasLotSeq)

  var seedBPass = await scrypt.async(passPoint, Buffer.concat([addressHash, ownerEntropy]), 1024, 1, 1, 64, undefined, promiseInterval)

  return finishDecryptECMult(seedBPass, encryptedPart1, encryptedPart2, passInt, compressed)
}

function decryptECMult (buffer, passphrase, progressCallback, scryptParams) {
  buffer = buffer.slice(1) // FIXME: we can avoid this
  passphrase = Buffer.from(passphrase.normalize('NFC'), 'utf8')
  scryptParams = scryptParams || SCRYPT_PARAMS
  const {
    addressHash,
    encryptedPart1,
    encryptedPart2,
    ownerEntropy,
    ownerSalt,
    hasLotSeq,
    compressed,
    N,
    r,
    p
  } = prepareDecryptECMult(buffer, passphrase, progressCallback, scryptParams)
  var preFactor = scrypt(passphrase, ownerSalt, N, r, p, 32, progressCallback)

  const {
    passInt,
    passPoint
  } = getPassIntAndPoint(preFactor, ownerEntropy, hasLotSeq)

  var seedBPass = scrypt(passPoint, Buffer.concat([addressHash, ownerEntropy]), 1024, 1, 1, 64)

  return finishDecryptECMult(seedBPass, encryptedPart1, encryptedPart2, passInt, compressed)
}

function verify (string) {
  var decoded = bs58check.decodeUnsafe(string)
  if (!decoded) return false

  if (decoded.length !== 39) return false
  if (decoded.readUInt8(0) !== 0x01) return false

  var type = decoded.readUInt8(1)
  var flag = decoded.readUInt8(2)

  // encrypted WIF
  if (type === 0x42) {
    if (flag !== 0xc0 && flag !== 0xe0) return false

  // EC mult
  } else if (type === 0x43) {
    if ((flag & ~0x24)) return false
  } else {
    return false
  }

  return true
}

module.exports = {
  decrypt: decrypt,
  decryptECMult: decryptECMult,
  decryptRaw: decryptRaw,
  encrypt: encrypt,
  encryptRaw: encryptRaw,
  decryptAsync: decryptAsync,
  decryptECMultAsync: decryptECMultAsync,
  decryptRawAsync: decryptRawAsync,
  encryptAsync: encryptAsync,
  encryptRawAsync: encryptRawAsync,
  verify: verify
}


lib/bip38.js
var aes = require('browserify-aes')
var assert = require('assert')
var createHash = require('create-hash')
var cs = require('coinstring')
var scrypt = require('scryptsy')

var ecurve = require('ecurve')
var curve = ecurve.getCurveByName('secp256k1')

var BigInteger = require('bigi')

function bufferXOR (buf1, buf2) {
  assert.equal(buf1.length, buf2.length)

  var out = new Buffer(buf1.length)
  for (var i = 0; i < buf1.length; i++) {
    out[i] = buf1[i] ^ buf2[i]
  }

  return out
}

// SHA256(SHA256(buffer))
function sha256x2 (buffer) {
  buffer = createHash('sha256').update(buffer).digest()
  return createHash('sha256').update(buffer).digest()
}

function Bip38 (versions) {
  if (!(this instanceof Bip38)) return new Bip38()

  // default to Bitcoin WIF versions
  this.versions = versions || { private: 0x80 }

  // BIP38 recommended
  this.scryptParams = {
    N: 16384,
    r: 8,
    p: 8
  }
}

Bip38.prototype.encryptRaw = function (buffer, compressed, passphrase, saltAddress) {
  assert.equal(buffer.length, 32, 'Invalid private key length')

  var secret = new Buffer(passphrase, 'utf8')
  var salt = sha256x2(saltAddress).slice(0, 4)

  var N = this.scryptParams.N
  var r = this.scryptParams.r
  var p = this.scryptParams.p

  var scryptBuf = scrypt(secret, salt, N, r, p, 64)
  var derivedHalf1 = scryptBuf.slice(0, 32)
  var derivedHalf2 = scryptBuf.slice(32, 64)

  var xorBuf = bufferXOR(buffer, derivedHalf1)
  var cipher = aes.createCipheriv('aes-256-ecb', derivedHalf2, new Buffer(0))
  cipher.setAutoPadding(false)
  cipher.end(xorBuf)

  var cipherText = cipher.read()

  // 0x01 + 0x42 + flagByte + salt + cipherText
  var flagByte = compressed ? 0xe0 : 0xc0
  var prefix = new Buffer(3)
  prefix.writeUInt8(0x01, 0)
  prefix.writeUInt8(0x42, 1)
  prefix.writeUInt8(flagByte, 2)

  return Buffer.concat([prefix, salt, cipherText])
}

Bip38.prototype.encrypt = function (wif, passphrase, saltAddress) {
  var d = cs.decode(wif).slice(1)
  var compressed = (d.length === 33) && (d[32] === 0x01)

  // truncate the compression flag
  if (compressed) {
    d = d.slice(0, -1)
  }

  return cs.encode(this.encryptRaw(d, compressed, passphrase, saltAddress))
}

// some of the techniques borrowed from: https://github.com/pointbiz/bitaddress.org
// todo: (optimization) init buffer in advance, and use copy instead of concat
Bip38.prototype.decryptRaw = function (encData, passphrase) {
  // 39 bytes: 2 bytes prefix, 37 bytes payload
  assert.equal(encData.length, 39, 'Invalid BIP38 data length')

  // first byte is always 0x01
  assert.equal(encData.readUInt8(0), 0x01, 'Invalid BIP38 prefix')

  // check if BIP38 EC multiply
  var type = encData.readUInt8(1)
  if (type === 0x43) {
    return this.decryptECMult(encData, passphrase)
  }

  passphrase = new Buffer(passphrase, 'utf8')

  assert.equal(type, 0x42, 'Invalid BIP38 type')
  var flagByte = encData.readUInt8(2)
  var compressed = flagByte === 0xe0

  if (!compressed) {
    assert.equal(flagByte, 0xc0, 'Invalid BIP38 compression flag')
  }

  var N = this.scryptParams.N
  var r = this.scryptParams.r
  var p = this.scryptParams.p

  var addresshash = encData.slice(3, 7)
  var scryptBuf = scrypt(passphrase, addresshash, N, r, p, 64)
  var derivedHalf1 = scryptBuf.slice(0, 32)
  var derivedHalf2 = scryptBuf.slice(32, 64)

  var privKeyBuf = encData.slice(7, 7 + 32)
  var decipher = aes.createDecipheriv('aes-256-ecb', derivedHalf2, new Buffer(0))
  decipher.setAutoPadding(false)
  decipher.end(privKeyBuf)

  var plainText = decipher.read()
  var privateKey = bufferXOR(plainText, derivedHalf1)

  return {
    privateKey: privateKey,
    compressed: compressed
  }
}

Bip38.prototype.decrypt = function (encryptedBase58, passphrase) {
  var encBuffer = cs.decode(encryptedBase58)
  var decrypt = this.decryptRaw(encBuffer, passphrase)

  // Convert to WIF
  var bufferLen = decrypt.compressed ? 34 : 33
  var buffer = new Buffer(bufferLen)

  buffer.writeUInt8(this.versions.private, 0)
  decrypt.privateKey.copy(buffer, 1)

  if (decrypt.compressed) {
    buffer.writeUInt8(0x01, 33)
  }

  return cs.encode(buffer)
}

Bip38.prototype.decryptECMult = function (encData, passphrase) {
  passphrase = new Buffer(passphrase, 'utf8')
  encData = encData.slice(1) // FIXME: we can avoid this

  var compressed = (encData[1] & 0x20) !== 0
  var hasLotSeq = (encData[1] & 0x04) !== 0

  assert.equal((encData[1] & 0x24), encData[1], 'Invalid private key.')

  var addresshash = encData.slice(2, 6)
  var ownerEntropy = encData.slice(6, 14)
  var ownerSalt

  // 4 bytes ownerSalt if 4 bytes lot/sequence
  if (hasLotSeq) {
    ownerSalt = ownerEntropy.slice(0, 4)

  // else, 8 bytes ownerSalt
  } else {
    ownerSalt = ownerEntropy
  }

  var encryptedPart1 = encData.slice(14, 22) // First 8 bytes
  var encryptedPart2 = encData.slice(22, 38) // 16 bytes

  var N = this.scryptParams.N
  var r = this.scryptParams.r
  var p = this.scryptParams.p
  var preFactor = scrypt(passphrase, ownerSalt, N, r, p, 32)

  var passFactor
  if (hasLotSeq) {
    var hashTarget = Buffer.concat([preFactor, ownerEntropy])
    passFactor = sha256x2(hashTarget)

  } else {
    passFactor = preFactor
  }

  var passInt = BigInteger.fromBuffer(passFactor)
  var passPoint = curve.G.multiply(passInt).getEncoded(true)

  var seedBPass = scrypt(passPoint, Buffer.concat([addresshash, ownerEntropy]), 1024, 1, 1, 64)
  var derivedHalf1 = seedBPass.slice(0, 32)
  var derivedHalf2 = seedBPass.slice(32, 64)

  var decipher = aes.createDecipheriv('aes-256-ecb', derivedHalf2, new Buffer(0))
  decipher.setAutoPadding(false)
  decipher.end(encryptedPart2)

  var decryptedPart2 = decipher.read()
  var tmp = bufferXOR(decryptedPart2, derivedHalf1.slice(16, 32))
  var seedBPart2 = tmp.slice(8, 16)

  var decipher2 = aes.createDecipheriv('aes-256-ecb', derivedHalf2, new Buffer(0))
  decipher2.setAutoPadding(false)
  decipher2.write(encryptedPart1) // first 8 bytes
  decipher2.end(tmp.slice(0, 8)) // last 8 bytes

  var seedBPart1 = bufferXOR(decipher2.read(), derivedHalf1.slice(0, 16))
  var seedB = Buffer.concat([seedBPart1, seedBPart2], 24)
  var factorB = sha256x2(seedB)

  // d = passFactor * factorB (mod n)
  var d = passInt.multiply(BigInteger.fromBuffer(factorB)).mod(curve.n)

  return {
    privateKey: d.toBuffer(32),
    compressed: compressed
  }
}

Bip38.prototype.verify = function (encryptedBase58) {
  var decoded
  try {
    decoded = cs.decode(encryptedBase58)
  } catch (e) {
    return false
  }

  if (decoded.length !== 39) return false
  if (decoded.readUInt8(0) !== 0x01) return false

  var type = decoded.readUInt8(1)
  var flag = decoded.readUInt8(2)

  // encrypted WIF
  if (type === 0x42) {
    if (flag !== 0xc0 && flag !== 0xe0) return false

  // EC mult
  } else if (type === 0x43) {
    if ((flag & ~0x24)) return false

  } else {
    return false
  }

  return true
}

module.exports = Bip38
